nemo/collections/nlp/data/common/sequence_to_sequence_dataset.py

```diff
@@ -146,76 +154,136 @@ def __init__(
+        logging.info(f'Source Dataset Length : {len(self.src_indexed_dataset)}')
+        logging.info(f'Target Dataset Length : {len(self.tgt_indexed_dataset)}')
 
     def __len__(self):
-        return len(self.src_dataset)
+        if self.max_num_samples is None:
+            return len(self.src_indexed_dataset)
+        else:
+            return self.max_num_samples
 
     def __getitem__(self, idx):
-        src = self.src_dataset[idx]
+        if isinstance(idx, np.int64):
+            idx = idx.item()
+
+        if self.samples_mapping is not None:
+            assert idx < len(self.samples_mapping)
+            idx, _, _ = self.samples_mapping[idx]
+            if isinstance(idx, np.uint32):
+                idx = idx.item()
+
+        assert idx < len(self.src_indexed_dataset)
+        src = self.src_indexed_dataset[idx]
         if len(src) > self.max_src_seq_length - 2:
             src = src[: self.max_src_seq_length - 2]
+        text_enc = np.concatenate([[self.src_tokenizer.bos_id], src, [self.src_tokenizer.eos_id]])
 
-        src = [self.src_tokenizer.bos_id] + src + [self.src_tokenizer.eos_id]
-        tgt = self.tgt_dataset[idx]
+        tgt = self.tgt_indexed_dataset[idx]
+        if len(tgt) > self.max_tgt_seq_length - 2:
+            tgt = tgt[: self.max_tgt_seq_length - 2]
 
-        if len(tgt) > self.max_tgt_seq_length - 1:
-            tgt = tgt[: self.max_tgt_seq_length - 1]
+        text_dec = np.concatenate([[self.tgt_tokenizer.bos_id], tgt])
+        labels = np.concatenate([tgt, [self.tgt_tokenizer.eos_id]])
 
-        text_enc = src
-        text_dec = [self.tgt_tokenizer.bos_id] + tgt
-        labels = tgt + [self.tgt_tokenizer.eos_id]
         return {'text_enc': text_enc, 'text_dec': text_dec, 'labels': labels}
 
+    def _build_samples_mapping(self):
+        if self.max_num_samples is not None:
+            # This means max src and max tgt sequence length need to be the same
+            if self.max_src_seq_length != self.max_tgt_seq_length:
+                raise ValueError(
+                    f"max_src_seq_length ({self.max_src_seq_length}) != max_tgt_seq_length ({self.max_tgt_seq_length}). This is needed for max_samples based training for now."
+                )
+
+            self.samples_mapping = get_samples_mapping(
+                indexed_dataset=self.src_indexed_dataset,
+                data_prefix=self.src_file_name,
+                num_epochs=None,
+                max_num_samples=self.max_num_samples,
+                max_seq_length=self.max_src_seq_length - 2,
+                short_seq_prob=0,
+                seed=self.seed,
+                name=self.src_file_name.split('/')[-1],
+                binary_head=False,
+            )
+        else:
+            self.samples_mapping = None
+
+
+class TextMemmapSequenceToSequenceDataset(IndexedSequenceToSequenceDataset):
+    """Memory-mapped text sequence to sequence dataset. Operates on raw text files and tokenizes the text on-the-fly."""
```

Comment:
Is there exactly one source file and one target file?

Reply:
Please expand on the doc string here.

Reply:
Done. Added a docstring to the constructor that explains each input arg to the class.